File: python/xsCloneContext.py
#
# (c) 2012-2014 Scianta Analytics LLC   All Rights Reserved.  
# Reproduction or unauthorized use is prohibited. Unauthorized
# use is illegal. Violators will be prosecuted. This software 
# contains proprietary trade and business secrets.            
#
import sys, subprocess, os, platform, shutil
import splunk.Intersplunk as si

if __name__ == '__main__':

    new=''
    newScope='private'
    old=''    
    oldScope='private'
    inKeyword=''
    toKeyword=''
    try:
        if len(sys.argv) > 1:
            for arg in sys.argv[1:]:
                if arg.lower() == "to":
                    toKeyword="to"
                elif arg.lower() == "in":
                    inKeyword="in"
                elif toKeyword == '':
                    if inKeyword == '':
                        old=arg
                    else:
                        oldScope=arg
                        inKeyword = ''
                else:
                    if inKeyword == '':
                        new=arg
                    else:
                        newScope=arg
        else:
            raise Exception("xsCloneContext-F-001: Usage: xsCloneContext oldContext [IN scope] TO newContext [IN scope]");

        binary = os.environ["SPLUNK_HOME"] + "/etc/apps/xtreme/bin/" +  platform.system() + "/" + platform.architecture()[0] + "/xsCloneContext"
        if (platform.system() == 'Windows'):
            binary = binary + ".exe"
        if not os.path.isfile(binary):
            raise Exception("xsCloneContext-F-000: Can't find binary file " + binary)

        subprocess.call([binary, '-N', newScope, '-O', oldScope, '-n', new, '-o', old])

        if platform.system() == 'Windows':
            sys.stdout.flush()
            time.sleep(0.5)
    
    except Exception, e:
        si.generateErrorResults(e)

File: xsCloneContext.c
/*
 Copyright 2012-2014 Scianta Analytics LLC   All Rights Reserved.  
 Reproduction or unauthorized use is prohibited. Unauthorized
 use is illegal. Violators will be prosecuted. This software 
 contains proprietary trade and business secrets.            

 Program: xsCloneContext

 Usage: xsCloneContext -o oldContext [-O oldScope] -n newContext [-N newScope]
    -N scope of new context
    -O scope of old context
    -n new context
    -o old context

 Description:
    Makes a full copy of a Context, including all concepts.
*/
#include <errno.h>
#include <libgen.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "saContext.h"
#include "saLicensing.h"
#include "saSignal.h"
#include "saSplunk.h"

extern char *optarg;
extern int optind, optopt;

extern saContextTypePtr saSplunkContextLoad(char *, char *, int *, char *, char *);
extern bool saSplunkContextSave(saContextTypePtr, char *, int, char *, char *);
extern char *saSplunkGetRoot(char *);
extern int saSplunkGetScope(char *);
extern saSplunkInfoPtr saSplunkLoadHeader();
extern bool saSplunkReadInfoPathFile(saSplunkInfoPtr);

int main(int argc, char* argv[]) 
{
    saContextTypePtr contextPtr;
    bool argError;
    char newName[256];
    char oldName[256];
    char *root = saSplunkGetRoot(argv[0]);
    int c;
    int newScope = saSplunkGetScope(NULL);
    int oldScope = saSplunkGetScope(NULL);

    if (!isLicensed())
        exit(EXIT_FAILURE);

    initSignalHandler(basename(argv[0]));
    argError = false;
    newName[0] ='\0';
    oldName[0] ='\0';
    while ((c = getopt(argc, argv, "N:O:n:o:")) != -1) 
    {
        switch (c)
        {
            case 'n':
	        strcpy(newName, optarg);
                break;
            case 'N':
                newScope = saSplunkGetScope(optarg);
                break;
            case 'o':
                strcpy(oldName, optarg);
                break;
            case 'O':
                oldScope = saSplunkGetScope(optarg);
                break;
            case '?':
                fprintf(stderr, "xsCloneContext-F-101: Unrecognised option: -%c\n", optopt);
                argError = true;
        }
    }
    if (argError)
    {
        fprintf(stderr, "xsCloneContext-F-103: Usage: xsCloneContext -N newScope -O oldScope -n newContext -o oldContext");
        exit(EXIT_FAILURE);
    }

    saSplunkInfoPtr p = saSplunkLoadHeader();
    if (p == NULL)
    {
        fprintf(stderr, "xsCloneContext-F-105: Can't get info header\n");
        exit(EXIT_FAILURE);
    }

    if (saSplunkReadInfoPathFile(p) == false)
    {
        fprintf(stderr, "xsCloneContext-F-107: Can't read search results file %s\n",
                p->infoPath == NULL ? "NULL" : p->infoPath);
        exit(EXIT_FAILURE);
    }

    contextPtr = saSplunkContextLoad(oldName, root, &oldScope, p->app, p->user);
    if (contextPtr == NULL)
    {
        fprintf(stderr, "xsCloneContext-F-109: Can't load context: %s\n", oldName);
        exit(EXIT_FAILURE);
    }

    strcpy(contextPtr->name, newName);
    if (saSplunkContextSave(contextPtr, root, newScope, p->app, p->user) == false)
    {
        fprintf(stderr, "xsCloneContext-F-111: Can't save context: %s\n", newName);
        exit(EXIT_FAILURE);
    }
    exit(0);
}
